c2rust-refactor/scripts/upgrade_ptr_to_ref.lua

```diff
@@ -722,34 +978,56 @@ end
 -- so we don't accidentally access old info
 -- NOTE: If this script encounters any nested functions, this will reset variables
 -- prematurely. We should push and pop a stack of variable scopes to account for this
-function Visitor:visit_fn_decl(fn_decl)
-    self.vars = {}
+function Visitor:clear_nonstatic_vars()
+    local static_vars = {}
+
+    for hirid, var in pairs(self.vars) do
+        if var.kind == "static" then
+            static_vars[hirid] = var
+        end
+    end
+
+    self.vars = static_vars
 end
 
 function Visitor:flat_map_item(item, walk)
     local item_kind = item:get_kind()
 
     if item_kind == "Struct" then
         local lifetimes = OrderedMap()
-        local field_ids = item:get_field_ids()
+        local fields = item:get_fields()
+        local is_copy = true
 
-        for _, field_id in ipairs(field_ids) do
-            local ref_cfg = self.node_id_cfgs[field_id]
+        for _, field in ipairs(fields) do
+            local field_id = field:get_id()
+            local cfg = self.node_id_cfgs[field_id]
             local field_hrid = self.tctx:nodeid_to_hirid(field_id)
 
             self:add_field(field_hrid, Field.new(field_id))
 
-            if ref_cfg and ref_cfg.extra_data.lifetime then
-                item:add_lifetime(ref_cfg.extra_data.lifetime)
+            if cfg then
+                if cfg:is_box_any() then
+                    is_copy = false
+                end
+
+                if cfg.extra_data.lifetime then
+                    item:add_lifetime(cfg.extra_data.lifetime)
 
-                lifetimes[ref_cfg.extra_data.lifetime] = true
+                    lifetimes[cfg.extra_data.lifetime] = true
+                end
             end
         end
 
+        if not is_copy then
+            item:remove_copy_derive()
+        end
+
         local hirid = self.tctx:nodeid_to_hirid(item:get_id())
 
-        self:add_struct(hirid, Struct.new(lifetimes))
+        self:add_struct(hirid, Struct.new(lifetimes, is_copy))
     elseif item_kind == "Fn" then
+        self:clear_nonstatic_vars()
```

Comment:
Correct, and I ran into that problem when I rewrote some code into a simple closure (since they're functions too). So I tweaked it here a bit. So it's a little better but still likely doesn't handle all cases